_posts/2017-05-30-database-encryption.markdown

```diff
@@ -0,0 +1,172 @@
+```
+rails g symmetric_encryption:heroku_config
+```
+This will create `config/symmetric-encryption.yml` file and also output commands you can run to set _encrypted_ encryption key on Heroku as an environment variable.
+
+
+For someone in order to be able to decode the data:
+
+
+- They have to be able to access our database.
+- They need to be able to access our source code.
+- They need to be able to access key files on disk or Heroku's configuration.
+
+### ActiveRecord Integration
+Symmetric Encryption provides a seamless integration with `ActiveRecord`, we can use `attr_encrypted` helper method to define _encrypted_ fields. Lets say we wanted to encrypt a `Note` model that has `note` and `subject`, you can add following to your `ActiveRecord` model:
+
+```ruby
+# app/models/note.rb
+attr_encrypted :note
+attr_encrypted :subject
+```
+
+This means whenever you set `note` for this model `symmetric-encryption` will set `encrypted_note` field in the database. Whenever you retrieve an instance of this model, `symmetric-encryption` will decrypt `encrypted_note` field and you can access _decrypted_ value by just accessing `note`.
+
+In our case we couldn't use this helper immediately, the issue is, using `encrypted_attr` means we will no longer be able to use un-encrypted fields. In our migration process until 4th step we still use non-encrypted fields and we need to populate them. In order to do that we started with adding a `before_validation` callback to our model to set encrypted fields based on un-encrypted ones.
+
+```ruby
+# app/models/note.rb
+before_validation :encrypt_notes_fields
+
+def encrypt_note_fields
+  self.encrypted_note = SymmetricEncryption.encrypt(note, true, true, :string)
+  self.encrypted_subject = SymmetricEncryption.encrypt(subject, true, true, :string)
+end
+
+validates :encrypted_note, symmetric_encryption: true, if: :note?
+validates :encrypted_subject, symmetric_encryption: true, if: :subject?
+```
+In the above code `SymmetricEncryption.encrypt(note, true, true, :string)` means encrypt `note` field, use random IV, compress the string and also use string when decrypting.
+
+Those two `validates` methods above make sure that we don't allow setting un-encrypted values for those encrypted fields.
+
+Once we got the the 4th step and stopped populating/reading un-encrypted fields we can easily switch above to
+
+```ruby
+# app/models/note.rb
+attr_encrypted :note
+attr_encrypted :subject
+```
+
+# Query encrypted fields
+Generally when we encrypt a field we can't do a partial query on the content of that field. On the other hand if we use the same IV ([Initialization Vector](https://en.wikipedia.org/wiki/Initialization_vector)) each time we encrypt a value, we can do a exact match query. Using same IV means encrypting the same value always end up with the same encrypted string. If exact match query is not something you need, the recommended approach is to use random IV for each encryption.
```

Comment:
reverse IV and initialization vector, so _we use the same initialization vector (IV)_